### PR TITLE
build: improve cuda detection

### PR DESCRIPTION
**Pull request by iblis17**
- change ADD_LDFLAGS
        - https://github.com/apache/incubator-mxnet/pull/7856
        - https://github.com/apache/incubator-mxnet/pull/7838

- guessing the CUDA_HOME if user not provides it.

----

@SimonEnsemble
Could you try this patch out with a clean clone?

### CODE REVIEW

**Comment by SimonEnsemble** [comment]
^ It looks like some checks were not successful?

How can I checkout this particular version?
`Pkg.checkout("MXNet", "iblis17:issue-286")` does not work.

**Comment by vchuravy** [comment]
@SimonEnsemble you need to add as a remote https://github.com/iblis17/MXnet.jl or use Github `hub`

**Comment by SimonEnsemble** [comment]
I tried:
```
Pkg.clone("git://github.com/iblis17/MXnet.jl.git", "issue-286")
Pkg.build("issue-286")
```
and I still get:
```
LoadError: failed process: Process(`make -j8 USE_BLAS=openblas 'MSHADOW_LDFLAGS=-lm /home/cokes/juliav0.6/bin/../lib/julia/libopenblas64_.so'`, ProcessExited(2)) [2]
```
I do have CUDA installed for sure...
`fatal error: cuda.h: No such file or directory`

**Comment by iblis17** [comment]
Curious about where is your cuda and cudnn installation.
On my Linux box, it's /opt/cuda.

**Comment by iblis17** [comment]
Oh, 
you miss this step:
```
Pkg.checkout("issue-286", "issue-286")
```

Pkg.clone only rename the dir and checkout master for you,
so we need to manully checkout this brach of PR.
```

**Comment by SimonEnsemble** [comment]
In my `~/.bashrc`:
```
# for CUDA obviously
export PATH=/usr/local/cuda-8.0/bin:${PATH:+:${PATH}}
export LD_LIBRARY_PATH=/usr/local/cuda-8.0/lib64:${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
export LD_LIBRARY_PATH=/home/cokes/cudaNN/lib64:${LD_LIBRARY_PATH}
```
Which should I put as my `CUDA_HOME`? Is cudann required for MXNet too?

File: deps/build.jl [hunk 1]
```diff
@@ -53,18 +52,18 @@ end
 if HAS_CUDA  # then check cudnn
   let cudnnlib = Libdl.find_library("libcudnn", CUDAPATHS)
     HAS_CUDNN = !isempty(cudnnlib) && Libdl.dlopen_e(cudnnlib) != C_NULL
-    if HAS_CUDNN
-      # TODO: do more version check?
-      cudnnver = dec(ccall((:cudnnGetVersion, cudnnlib), Csize_t, ()))
+    if HAS_CUDNN && !haskey(ENV, "CUDA_HOME")  # inference `CUDA_HOME`
+      ENV["CUDA_HOME"] = dirname(dirname(cudnnlib))
```
Comment (iblis17) [inline deps/build.jl]: YFI: I try to guessing your CUDA_HOME via libcudnn's path

Comment (SimonEnsemble) [reply deps/build.jl]: How about this for guessing the path to CUDA?
`filter(x->contains(x, "cuda"), split(ENV["PATH"], ":"))`
I assume everyone with CUDA installed put it in the PATH variable...

**Comment by SimonEnsemble** [comment]
Starting from scratch, I:
```
Pkg.clone("git://github.com/iblis17/MXnet.jl.git", "issue-286")
Pkg.checkout("issue-286", "issue-286")
Pkg.build("issue-286")
```

but it is stuck... unless it takes >15 min to compile...

File: deps/build.jl (hunk 1 above)
Comment (iblis17) [reply deps/build.jl]: 🤔 , it does not set `PATH` and `LD_LIBRARY_PATH` on my machine, but works fine.

BTW, where is your libcublas
```
julia> Libdl.dlpath("libcublas")                                                                                                                  
"/opt/cuda/lib64/libcublas.so
```
seesm you have different installation path for CUDA and cuDNN,
maybe I should determin path from cuBLAS?

**Comment by iblis17** [comment]
> but it is stuck... unless it takes >15 min to compile...

yeah, on Travis CI, It takes > 20 min

**Comment by iblis17** [comment]
You can check if something eat up your CPUs via `top` or not.

**Comment by iblis17** [comment]
> Which should I put as my CUDA_HOME? Is cudann required for MXNet too?

* In your case, I expect `CUDA_HOME` is `/usr/local/cuda-8.0/`, `CUDA_HOME` should contains `include`, `bin` and `lib64` in it.
* cuDNN is optional.

**Comment by SimonEnsemble** [comment]
@iblis17 It compiles and the example on the README finally runs without errors!

So please do merge into main MXNet.
Another alternative is to just have users install MXNet from here:
https://mxnet.incubator.apache.org/get_started/install.html
and then always use
export MXNET_HOME before installation.
But when I do that, while it compiles, there are errors that arise in the examples, so it seems that MXNet.jl is predicated upon a different version of MXNet.

**Comment by iblis17** [comment]
🎉 

Let's wait for @vchuravy 's review.

>Another alternative is to just have users install MXNet from here:
https://mxnet.incubator.apache.org/get_started/install.html
and then always use
export MXNET_HOME before installation.
But when I do that, while it compiles, there are errors that arise in the examples, so it seems that MXNet.jl is predicated upon a different version of MXNet.

Master branch of MXNet.jl requires master branch of upstream mxnet.
Maybe checkout the stable release MXNet.jl can works with the prebuilt binary version. (but I don't try it yet)

hmm ... https://github.com/dmlc/MXNet.jl/releases
the last release is for upstream mxnet 0.10.0 😱 ...

**Comment by SimonEnsemble** [comment]
Now I'm installing MXNet.jl on CentOS.
I have a working cuda installation in `/usr/local/apps/cuda/cuda-8.0`.

On master (via `Pkg.checkout("MXNet")`), upon `Pkg.build("MXNet")`, it does not detect the CUDA installation.

I then:
`ENV["CUDA_HOME"] = "/usr/local/apps/cuda/cuda-8.0"; Pkg.build("MXNet")`
and still get the message:
`INFO: Did not find a CUDA installation, using CPU-only version of MXNet.`

What can I do to troubleshoot?

**Comment by pluskid** [comment]
@SimonEnsemble have you tried to use [manual installation](http://dmlc.ml/MXNet.jl/latest/user-guide/install/#Manual-Compilation-1) that compile libmxnet separately and point to the compiled library?

**Comment by CorySimon** [comment]
@pluskid Yes, I've successfully built MXNet.jl pointing to a version of MXNet that I compiled myself. But, then the examples in MXNet.jl do not even run... See here:
https://github.com/dmlc/MXNet.jl/issues/286

**Pull request merged**